FlowCrypt/Functionality/Services/GoogleUserService.swift

```diff
@@ -70,67 +69,61 @@ extension GoogleUserService: UserServiceType {
-                presenting: viewController
-            ) { authState, error in
-                if let authState = authState {
-                    let missingScopes = self.checkMissingScopes(authState.scope)
-                    if !missingScopes.isEmpty {
-                        reject(GoogleUserServiceError.userNotAllowedAllNeededScopes(missingScopes: missingScopes))
-                        return
-                    }
-
-                    let authorization = GTMAppAuthFetcherAuthorization(authState: authState)
-                    guard let email = authorization.userEmail else {
-                        reject(GoogleUserServiceError.inconsistentState("Missed email"))
-                        return
-                    }
-
-                    self.saveAuth(state: authState, for: email)
-
-                    guard let token = authState.lastTokenResponse?.accessToken else {
-                        reject(GoogleUserServiceError.inconsistentState("Missed token"))
-                        return
-                    }
-
-                    self.fetchGoogleUser(with: authorization) { result in
-                        switch result {
-                        case .success(let user):
-                            resolve(SessionType.google(email, name: user.name, token: token))
-                        case .failure(let error):
-                            self.handleUserInfo(error: error)
-                            reject(error)
+    func signIn(in viewController: UIViewController) async throws -> SessionType {
+        return try await withCheckedThrowingContinuation { continuation in
+            DispatchQueue.main.async {
+                // todo - should be fixed with MainActor instead?
+                // Google doesn't like to be called on non-main thread
+                let request = self.makeAuthorizationRequest()
+                let googleAuthSession = OIDAuthState.authState(
+                    byPresenting: request,
+                    presenting: viewController
+                ) { authState, error in
+                    if let authState = authState {
+                        Task<Void, Never> {
+                            do {
+                                return continuation.resume(returning: try await self.handleGoogleAuthStateResult(authState))
+                            } catch {
+                                return continuation.resume(throwing: error)
+                            }
                         }
+                        return
                     }
-                } else if let error = error {
-                    reject(error)
-                } else {
-                    fatalError("Shouldn't happe because covered received non nil error and non nil authState")
+                    return continuation.resume(throwing: error ?? AppErr.unexpected("Shouldn't happen because covered received non nil error and non nil authState"))
+                }
+                DispatchQueue.main.async {
```

Comment:
It's `DispatchQueue.main.async` inside `DispatchQueue.main.async`, probably should work with just parent one.

Reply:
Well spotted 👍